sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -335,12 +343,30 @@ abstract class TypeCoercionBase {
       // Return the result after the widen data types have been found for all the children
       if (attrIndex >= children.head.output.length) return castedTypes.toSeq
 
-      // For the attrIndex-th attribute, find the widest type
-      val widenTypeOpt = findWiderCommonType(children.map(_.output(attrIndex).dataType))
-      castedTypes.enqueue(widenTypeOpt)
+      val outputType = findOutputType(children, attrIndex)
+      castedTypes.enqueue(outputType)
       getWidestTypes(children, attrIndex + 1, castedTypes)
     }
 
+    /** Given children of the operator, determines the output type of the
+     * `attrIndex`-th attribute
+     */
+    private def findOutputType(children: Seq[LogicalPlan], attrIndex: Int): Option[DataType] = {
```

Comment:
I find this weird. So far we always special cased expressions in CollationTypeCast. As @cloud-fan we have understanding that collations are only touched in that rule. This moves us to making double work. I would say it is better to move the widening of collations to CollationTypeCast by special casing the setOperations there (except, union, intersect) it is only three expressions. This is because we keep code in a cleaner state in that way. What do you guys think? @stefankandic @vladanvasi-db

Reply:
Also, who is to guarantee that we only have strings here, your change assumes that all attrIndexed at the same place will have same type, why do we then even call widening if that is the case?

Reply:
By same type I mean strings but with different collations.

Reply:
You cannot catch this in `CollationTypeCasts` or any type coercion rule for that matter, since they only operate on individual expressions, and here the projects are spread across different subqueries

Reply:
Could we then just check Unpivot as well? There are three rules that are run before CollationTypeCasts and those are 
```
UnpivotCoercion ::
WidenSetOperationTypes ::
ProcedureArgumentCoercion
```

Reply:
And maybe adding a comment in the typeCoercionRules should be added to give attention to string collations when adding new rules.